lambdas/scripts/dynamodb_migration_prmp_198.py

```diff
@@ -0,0 +1,160 @@
+
+    def process_entries(
+        self, label: str, all_entries: list[dict], total_count: int, update_fn
+    ):
+        self.logger.info(f"Running {label} migration")
+
+        for index, entry in enumerate(all_entries, start=1):
+            item_id = entry.get("ID")
+            self.logger.info(
+                f"[{label}] Processing item {index} of {total_count} (ID: {item_id})"
+            )
+
+            updated_fields = update_fn(entry)
+            if not updated_fields:
+                self.logger.debug(
+                    f"[{label}] Item {item_id} does not require update, skipping."
+                )
+                continue
+
+            if self.dry_run:
+                self.logger.info(
+                    f"[Dry Run] Would update item {item_id} with {updated_fields}"
+                )
+            else:
+                self.logger.info(f"Updating item {item_id} with {updated_fields}")
+                self.dynamo_service.update_item(
+                    table_name=self.target_table,
+                    key_pair={"ID": item_id},
+                    updated_fields=updated_fields,
+                )
+
+        self.logger.info(f"{label} migration completed.")
+
+    # def run_version_migration(self, all_entries, total_count):
+    #     self.process_entries(
+    #         "Version",
+    #         all_entries,
+    #         total_count,
+    #         lambda e: {"Version": 1} if e.get("Version") != 1 else None,
+    #     )
+
+    def run_author_migration(self, all_entries: list[dict], total_count: int) -> None:
+        self.logger.info("Running author migration")
+
+        bulk_reports = self.dynamo_service.scan_whole_table(self.bulk_upload_table)
+
+        bulk_upload_report_lookup: dict[str, dict] = {}
+
+        for row in bulk_reports:
+            nhs = row.get("NhsNumber")
+            status = row.get("UploadStatus")
+            timestamp = row.get("Timestamp")
+
+            if not nhs or status != "complete" or not timestamp:
+                continue
+
+            stored_report = bulk_upload_report_lookup.get(nhs)
+            if not stored_report or int(timestamp) > int(
+                stored_report.get("Timestamp", 0)
+            ):
+                bulk_upload_report_lookup[nhs] = row
+
+        def author_update(entry: dict) -> dict | None:
+            current_author = entry.get("Author")
+            nhs_number = entry.get("NhsNumber")
+
+            if current_author:
+                return None
+
+            bulk_upload_row = bulk_upload_report_lookup.get(nhs_number)
+            if not bulk_upload_row:
+                self.logger.warning(
+                    f"No matching completed bulk upload found for NHS {nhs_number}"
+                )
+                return None
+
+            new_author = bulk_upload_row.get("UploaderOdsCode")
+            if not new_author:
+                self.logger.warning(f"No uploader ODS code found for NHS {nhs_number}")
+                return None
```

Comment:
I think we might need to have a think about what ends up on the DLQ and what we skip.
If we don't have the data and never feasibly get it back we have to skip?